lib/src/js/js.dart

```diff
@@ -0,0 +1,80 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'package:js/js.dart';
+import 'package:js/js_util.dart';
+import 'package:node_interop/process.dart';
+
+@JS('process')
+external final Process? _process; // process is undefined in the browser
+
+@JS('history')
+external final Object? _historyApi; // history is undefined in Node.JS
+
+/// This extension adds `maybe<Property>` getters that return non-nullable
+/// properties with a nullable type.
+extension _PartialProcess on Process {
+  /// Returns [release] as nullable.
+  Release? get _maybeRelease => release;
+}
+
+/// Whether this Dart code is running in a strict mode context.
+///
+/// In strict mode, assigning properties to primitive types is an error. This
+/// uses that fact to sniff whether strict mode is in effect.
+final _isStrictMode = () {
+  try {
+    setProperty('', 'name', null);
+    return false;
+  } catch (_) {
+    return true;
+  }
+}();
+
+const bool isJS = true;
+
+bool get isNodeJs => _process?._maybeRelease?.name == 'node';
+
+bool get isBrowser =>
+    !isNodeJs &&
+    _historyApi != null &&
+    // Checking the type at runtime with `is` avoids throwing an error in the
+    // unlikely case that the returned object isn't a `string`.
+    getProperty<Object?>(_historyApi!, 'scrollRestoration') is String;
```

Comment:
Probably better to look for an API that's a little less recent, in case this ends up running on an older browser. If `window.location.href` is polyfilled, maybe check for the top-level `document` object?

Reply:
fair, I still want to check for a property rather than just the object so I replaced this by checking for document.querySelector, that property isn't likely to be polyfilled I hope 😅